### ISSUE TITLE
CITR Phase 1: Aggregate Unit Test Logs

### ISSUE THREAD

**Issue by rbarkerSL** [hashgraph/hedera-services#16729]
- Collect unit test artifacts by the build application workflow on push to develop and store as build artifacts
- Pull in unit test artifacts into XTS run
- Collect unit test artifacts on XTS run
- Aggregate test artifacts and store as build artifacts on XTS
- Publish all artifacts (commit merged by pull request AND XTS to codacy)

**Comment by rbarkerSL** [comment]
Use the workflow built in #16662 to collect up the logs

**Comment by rbarkerSL** [comment]
Additional clarifications from our meeting with Nathan last week:

1. We need to rename the `Test Reports` artifact and the `Unit Test Reports` artifact to be more descriptive related to their content.
  a. `Unit Test Reports` contains all of the XML reports that we need to aggregate together. We might consider naming this report `XML Test Report Data`.
  b. `Test Reports` contains html representations of the test report data that are human readable. We should rename this artifact to `Unit Test Reports`
2. We need to aggregate the test report info into a single test report artifact that can be collected by the downstream job
  a. Run the XML Test Report Data files through the `Publish Unit Test Result Action` to gain an aggregated test report for codecov/codacy
  b. Save this as a new `Aggregated XML Test Report` artifact
  c. In the XTS workflow pull in the `Aggregated XML Test Report` and the `XML Test Report Data` artifacts
  d. Run through the XTS workflow
  e. Capture the `XTS XML Test Report Data` artifact (new artifact)
  f. Run the `XTS XML Test Report Data` through the `Publish Unit Test Result Action` to generate an aggregated test report. Capture this as an artifact
  g. Combine the `XTS XML Test Report Data` with the  `XML Test Report Data` pulled in from the upstream build and run that through `Publish Unit Test Result Action` to generate the `Combined Test Report` artifact. I believe this is the piece that actually gets uploaded.

**Comment by mishomihov00** [comment]
@rbarkerSL Something is off. The `Publish Unit Test Result Action` generates a JSON output which seems to be used only by the action itself and could not be uploaded to Codacy - https://github.com/step-security/publish-unit-test-result-action/tree/cc82caac074385ae176d39d2d143ad05e1130b2d/?tab=readme-ov-file#json-result

**Comment by mishomihov00** [comment]
@rbarkerSL Also both the `Node: Build Application` and `ZXCron: Extended Test Suite` workflows are calling the `ZXC: [Node] Compile Application Code` workflow which actually generates the artifacts `Test Reports` and `Unit Test Reports`. Whenever this workflow is called with the `enable-unit-tests` flag set to `true` it not only runs the Unit tests but also starts an `aggregation` gradle job ([Check here](https://github.com/hashgraph/hedera-services/blob/e20b66ae32012a018bf82ae4913966966ce2a00c/.github/workflows/node-zxc-compile-application-code.yaml#L209)) and also uploads the results to codecov and codacy ([Check here](https://github.com/hashgraph/hedera-services/blob/e20b66ae32012a018bf82ae4913966966ce2a00c/.github/workflows/node-zxc-compile-application-code.yaml#L543)). I have the feeling that what we are trying to do is already done, but of course I might be wrong.

**Comment by rbarkerSL** [comment]
> @rbarkerSL Also both the `Node: Build Application` and `ZXCron: Extended Test Suite` workflows are calling the `ZXC: [Node] Compile Application Code` workflow which actually generates the artifacts `Test Reports` and `Unit Test Reports`. Whenever this workflow is called with the `enable-unit-tests` flag set to `true` it not only runs the Unit tests but also starts an `aggregation` gradle job ([Check here](https://github.com/hashgraph/hedera-services/blob/e20b66ae32012a018bf82ae4913966966ce2a00c/.github/workflows/node-zxc-compile-application-code.yaml#L209)) and also uploads the results to codecov and codacy ([Check here](https://github.com/hashgraph/hedera-services/blob/e20b66ae32012a018bf82ae4913966966ce2a00c/.github/workflows/node-zxc-compile-application-code.yaml#L543)). I have the feeling that what we are trying to do is already done, but of course I might be wrong.

Yes they both do this independently for their respective jobs. The goal here is to combine them into a single artifact. That way a user would have access to the data from either job independently AND from a combined set if they wanted the full picture.

@nathanklick please weigh in if I'm missing something here.

**Comment by mishomihov00** [comment]
A little note to my previous comment - Currently this `aggregation` gradle job is started only when the `ZXC: [Node] Compile Application Code` workflow is called from within the `Node: Build Application` workflow. So basically no `jacoco/testCodeCoverageReport/testCodeCoverageReport.xml` file is generated from XTS.

What I've managed to achieve now is to generate such file (`testCodeCoverageReport.xml`) also for the cases in which the `ZXC: [Node] Compile Application Code` workflow is called from within the `ZXCron: Extended Test Suite` workflow. This additional file should contain the aggregated results from the `Timing Sensitive Unit Tests`.

After that I'm combining this new file with the already generated one from the `ZXC: [Node] Compile Application Code` workflow execution from within the `Node: Build Application` workflow, using a python script (there might be a better solution for that) and I'm uploading this new file to codacy.

It seems that the upload is successful ([Check here](https://github.com/hashgraph/hedera-services/actions/runs/12372180277/job/34530003789#step:9:1)), but I'm not able to check what is actually being uploaded as I'm still waiting for access to the codacy `hashgraph` org. I guess that @nathanklick should provide access to there.
